File: 300.py
from ast import List
from bisect import bisect_left

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        lis = []
        for x in nums:
            idx = bisect_left(lis, x, 0, len(lis))
            if idx == len(lis):
                lis.append(x)
            else:
                lis[idx] = x
        return len(lis)

File: 300.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> LIS;
        for (auto i : nums) {
            auto iter = lower_bound(begin(LIS), end(LIS), i);
            if (iter == end(LIS)) LIS.push_back(i);
            else *iter = i;
        }
        return LIS.size();
    }
};